### ISSUE TITLE
Incorrect toolchain when opening a vsxmake solution using Visual Studio 2022

### ISSUE THREAD

**Issue by Zino2201** [xmake-io/xmake#1838]
### Describe the bug
When opening a solution generated using vsxmake (2022) with VS2022, it defaults to the Visual Studio 2010 toolchain. This does not happend with VS2019.

### Expected behavior
Visual Studio projects to use the VS2022 toolchain when generating a solution using VS2022.

### Related Environment

- xmake version: v2.5.9
- os: Windows 11 64-bit
- target platform: MSVC 2022 v143

**Comment by waruqi** [comment]
Can you upload your generated .sln .vcxproj files as zip?

**Comment by Zino2201** [comment]
> Can you upload your generated .sln .vcxproj files as zip?
[vsxmake2022.zip](https://github.com/xmake-io/xmake/files/7554315/vsxmake2022.zip)

**Comment by Zino2201** [comment]
Okay, so small update I uninstalled VS 2019 to check if for some reasons it create a compability problem with VS 2022 and it seems that the case because now xmake can't find VS2022 (before it did and showed "2022")

![xmake_vs2022_error01](https://user-images.githubusercontent.com/15857548/142211500-f1b29ec9-79e5-42ee-a1a7-0b899ae19204.png)

**Comment by waruqi** [comment]
> > Can you upload your generated .sln .vcxproj files as zip?
> 
> [vsxmake2022.zip](https://github.com/xmake-io/xmake/files/7554315/vsxmake2022.zip)

```
Microsoft Visual Studio Solution File, Format Version 12.0
# Visual Studio Version 17.0
#
# xmake version 2.5.9+HEAD.809ee9806
#
#   This file is auto generated by xmake.
#
#   Please DO NOT EDIT since all your modification will lost after re-generation.
#   See https://github.com/xmake-io/xmake/pull/472 for more infomation.
#
VisualStudioVersion = 17.0
MinimumVisualStudioVersion = 10.0
```

I found that your sln file has been vs2020 (VisualStudioVersion 17.0).

I didn't find any problems. It's ok.

**Comment by Zino2201** [comment]
> > > Can you upload your generated .sln .vcxproj files as zip?
> > 
> > 
> > [vsxmake2022.zip](https://github.com/xmake-io/xmake/files/7554315/vsxmake2022.zip)
> 
> ```
> Microsoft Visual Studio Solution File, Format Version 12.0
> # Visual Studio Version 17.0
> #
> # xmake version 2.5.9+HEAD.809ee9806
> #
> #   This file is auto generated by xmake.
> #
> #   Please DO NOT EDIT since all your modification will lost after re-generation.
> #   See https://github.com/xmake-io/xmake/pull/472 for more infomation.
> #
> VisualStudioVersion = 17.0
> MinimumVisualStudioVersion = 10.0
> ```
> 
> I found that your sln file has been vs2020 (VisualStudioVersion 17.0).
> 
> I didn't find any problems. It's ok.

Yeah it's seems good in the sln file but in VS it is set to vs2010:
![unknown](https://user-images.githubusercontent.com/15857548/142222072-442a991a-18a8-413b-9d72-60754a9e7b52.png)

**Comment by waruqi** [comment]
I only know that as long as the version in sln is correct, it should be correct. I don't know why vs will show this? Is the tool chain being used right now? If it is only a display problem, it should not affect normal compilation.

**Comment by Zino2201** [comment]
> I only know that as long as the version in sln is correct, it should be correct. I don't know why vs will show this? Is the tool chain being used right now? If it is only a display problem, it should not affect normal compilation.

Well I can't test right now as xmake now doesn't detect VS2022, even after a full reinstallation + restart. I'm using the release version of Visual Studio and not the preview. CMake detects it so I think it's not related to a env-var or something like that 🤔

![xmake_vs2022_error01](https://user-images.githubusercontent.com/15857548/142211500-f1b29ec9-79e5-42ee-a1a7-0b899ae19204.png)

**Comment by Zino2201** [comment]
Okay so I digged a bit more, VS2022 doesn't create set any environnement variables for some reason (VCInstallDir didn't exist, and VSInstallDir was still pointing my old VS2019 installation), it seems like CMake search VS using vswhere.exe that is always installed in `Program Files (x86)\Installer\vswhere.exe`, so maybe it is the new way of searching VS. For compilation you're right it doesn't affect, I forgot that xmake has its own custom targets. Thanks!

**Comment by waruqi** [comment]
>  I see xmake rely on environnement variables only. For compilation you're right it doesn't affect, I forgot that xmake has its own custom targets. Thanks!

no, xmake doesn't depend on vs envs. you can run xmake in clean cmd. It will use vswhere.

You can run `xmake l detect.sdks.find_vstudio` and let me see the output.

and please do not run xmake in vs prompt terminal. only need cmd.

**Comment by waruqi** [comment]
I don't have the environment to reproduce it now. You can create a project with vs2022 and compare the differences of vcxproj files

**Comment by SirLynix** [comment]
It seems to be coming from the `PlatformToolset`.

A regular project has this:
```
  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'" Label="Configuration">
    <ConfigurationType>Application</ConfigurationType>
    <UseDebugLibraries>true</UseDebugLibraries>
    <PlatformToolset>v143</PlatformToolset>
    <CharacterSet>Unicode</CharacterSet>
  </PropertyGroup>
```

But it doesn't appear in vcxproj generated for vsxmake

**Comment by SirLynix** [comment]
Ok I can confirm it comes from the PlatformToolset property.

When you fix this with VS2022, it adds this to xmake project:
```
  <PropertyGroup Label="Configuration" Condition="'$(Configuration)|$(Platform)'=='debug|x64'">
    <PlatformToolset>v143</PlatformToolset>
  </PropertyGroup>
  <PropertyGroup Label="Configuration" Condition="'$(Configuration)|$(Platform)'=='releasedbg|x64'">
    <PlatformToolset>v143</PlatformToolset>
  </PropertyGroup>
  <PropertyGroup Label="Configuration" Condition="'$(Configuration)|$(Platform)'=='coverage|x64'">
    <PlatformToolset>v143</PlatformToolset>
  </PropertyGroup>
  <PropertyGroup Label="Configuration" Condition="'$(Configuration)|$(Platform)'=='asan|x64'">
    <PlatformToolset>v143</PlatformToolset>
  </PropertyGroup>
```

**Comment by waruqi** [comment]
I try fix it. https://github.com/xmake-io/xmake/commit/c986baaa3b9f79d3add7b26e5ab8dd0cb75b735b

try it again?

**Comment by SirLynix** [comment]
It's fixed! Thank you.

**Comment by waruqi** [comment]
@Zino2201 Does it work for you?

**Comment by Zino2201** [comment]
@waruqi Just tried and yep it works now, thanks!
